### PR TITLE
Added fast paths in std.utf.byUTF

### PR DESCRIPTION
**Pull request by JackStouffer**
@9il @DmitryOlshansky

### CODE REVIEW

**Comment by 9il** [comment]
please benchmarks with combination with previous one PR

**Comment by JackStouffer** [comment]
Just tested it, still slower and with no noticeable difference in speed.

Hmm, I'm looking at the profile output and I don't see what could be causing the 8x slower speeds.

**Comment by 9il** [comment]
maybe inlining bug?

**Comment by JackStouffer** [comment]
I don't think so, but I am a novice at assembly. I found LDC's out put of the function, but I don't see it a call for it anywhere, so I guess it was inlined at every call point

File: std/utf.d [hunk 1]
```diff
@@ -3596,11 +3596,24 @@ template byUTF(C) if (isSomeChar!C)
...
+                    import std.ascii : isASCII;
+
                     if (pos == fill)
                     {
                         pos = 0;
-                        fill = cast(ushort)encode!(UseReplacementDchar.yes)(
-                            buf, decodeFront!(UseReplacementDchar.yes)(r));
+                        auto c = r.front;
+
+                        if (c.isASCII)
```
Comment (DmitryOlshansky) [inline std/utf.d]: Possibly isASCII is not inlined, just use plain old < 0x80

**Comment by JackStouffer** [comment]
Ok, I think I've figured it out. `byUTF` spends time decoding `dchar`s when `toUTF8` correctly skips that step with `dstring`s.

Will post numbers soon.

File: std/utf.d (hunk 1 above)
Comment (JackStouffer) [inline std/utf.d]: No effect

**Comment by JackStouffer** [comment]
Well, that did make the UTF string faster, but I forgot that the ascii string isn't using that code path, so no change :/

**Comment by JackStouffer** [comment]
Regardless of the other PR, these are two good optimizations that IMO should be pulled.

**Comment by JackStouffer** [comment]
Fixed test errors

**Comment by JackStouffer** [comment]
Ping

**Comment by JackStouffer** [comment]
Ping @9il

**Comment by wilzbach** [comment]
> please benchmarks with combination with previous one PR

How about starting to include such benchmarks in a special `benchmark` folder?
This would make it easier for future improvements & also help to prevent regressions.

I believe @burner is envisioning sth. similar with #2995, but we could still start to collect the benchmark script as (a) it comes "for free" (they are already written) and (b) might help to move forward with a future "full-blown" benchmark suite as then already have real scripts and know what functions / features are lacking.

Btw [runtime has already gone a [similar way](https://github.com/dlang/druntime/tree/master/benchmark).

(pinging @qznc here as he has written an [nice blog post](http://dlang.org/blog/2016/06/16/find-was-too-damn-slow-so-we-fixed-it/) about benchmarking Phobos)

**Comment by JackStouffer** [comment]
Ping @9il @klickverbot 

I think these two optimizations are good to go regardless of the other PR.

**Comment by JackStouffer** [comment]
@andralex Thanks!

**Pull request merged**